### ISSUE TITLE
Train on GPU and predict on CPU

### ISSUE THREAD

**Issue by stanbiryukov** [skorch-dev/skorch#553]
Hi there,

Is it possible to add functionality to allow training and predicting on different devices?  In basic_usage.ipynb, I tried training the model and then updating it as such:
```
net.set_params(device='cpu') # pred using CPU
y_pred = net.predict(X[:5])
# 'Expected object of device type cuda but got device type cpu for argument #2 'mat1' in call to _th_addmm'
```

The idea here is that predicting can then be done using multiple CPUs and may be faster than using the GPU.  I've done this successfully in pytorch by just changing the device before prediction.

**Comment by BenjaminBossan** [comment]
I haven't thought of that use case yet. What should probably work is dumping the model parameters and then loading them with `device='cpu'`, but there is probably a better way. Could you provide a minimal example in pure pytorch?

**Comment by stanbiryukov** [comment]
Here's a colab notebook containing the basic_usage.ipynb piece above, along with a pytorch regression example where fitting is done on GPU and prediction is done using CPU: https://colab.research.google.com/drive/1Fz8GJgGJOZFnwzrofPP1Phg-GYtXe9U-

I simply push the neural net model to cpu via .to('cpu') after training and then predict on a cpu tensor.

**Comment by BenjaminBossan** [comment]
> netcpu = copy.deepcopy(net).to('cpu')

Can you try that on the `module_` of the fitted `NeuralNetClassifier`? (sorry, can't test myself at the moment)

**Comment by stanbiryukov** [comment]
Confirmed this works
```
skcpumod = copy.deepcopy(net.module_).to('cpu')
skcpumod(torch.tensor(X).type(torch.FloatTensor).to('cpu'))
```

**Comment by BenjaminBossan** [comment]
Can you try setting the `module_` on the net and then using the net as usual?

```python
net.set_params(device='cpu')
skcpumod = copy.deepcopy(net.module_).to('cpu')
net.module_ = skcpumod
```

If this works, I wonder if we should consider adding a setter to `device` that does this under the hood.

**Comment by stanbiryukov** [comment]
Excellent, confirmed that the above works:
```
net = NeuralNetClassifier(
    ClassifierModule,
    max_epochs=20,
    lr=0.1,
    device='cuda'
)
net.fit(X, y)
y_pred_cuda = net.predict(X)

net.set_params(device='cpu')
skcpumod = copy.deepcopy(net.module_).to('cpu')
net.module_ = skcpumod
y_pred_cpu = net.predict(X)
np.array_equal(y_pred_cpu, y_pred_cuda) # True
```

**Comment by BenjaminBossan** [comment]
Great.

So now I wonder if it makes sense to add a setter to `device` or maybe add a `to` method to `NeuralNet` that basically performs this.

**Comment by stanbiryukov** [comment]
Defer to where you think is best.

**Comment by BenjaminBossan** [comment]
@ottonemo @thomasjpfan any opinion on that?

**Comment by thomasjpfan** [comment]
The use case here is to train on GPU and predict on CPU, which allows for https://github.com/skorch-dev/skorch/issues/553#issuecomment-551295775 to work.

If we were to have a `to` method, do we also update the references in the optimizer to the module on the cpu?

**Comment by stanbiryukov** [comment]
I think you'd want to convert all relevant objects of the class to CPU and be able to do the same back to GPU.  That way, if we need to keep training, we wouldn't run into issues where objects are on different devices.

**Comment by BenjaminBossan** [comment]
I think the issue with the references is an important one. We would need to make sure that this works, otherwise there could be subtle bugs.

**Comment by stanbiryukov** [comment]
Any additional thoughts to get this off the ground? I wonder if it would be best to add a 'def to_device' type function in utils.py that checks the device of an object and tries to cast it to either cpu or cuda.  Then the method in NeuralNet can loop through and call 'to_device'?

**Comment by githubnemo** [comment]
> Any additional thoughts to get this off the ground? I wonder if it would be best to add a 'def to_device' type function in utils.py that checks the device of an object and tries to cast it to either cpu or cuda. Then the method in NeuralNet can loop through and call 'to_device'?

We already have this addition in one of the open PRs. To me the bigger question is the interface. On one hand a `to` method on the neural net instance would be consistent with PyTorch and most importantly doesn't clash semantically with sklearn's `set_params`. On the other hand there is the existing framework of `set_params` which, for `lr` and its derivatives, already supports changes at runtime. In essence:

1. `net.to('cuda')`
OR
2. `net.set_params(device='cuda')`

I am unsure whether the latter is the best way to go forward, since the consequence of the call depends on the value that is modified. This might be confusing.

**Comment by stanbiryukov** [comment]
I see - either option sounds reasonable to me. Regarding the sklearn approach I don't think one would say grid search over that parameter so it doesn't have to be exposed. 

> since the consequence of the call depends on the value that is modified

Can you expand on that in reference to the second option?  How would that be different from any other set_params call?

**Comment by BenjaminBossan** [comment]
> We already have this addition in one of the open PRs

Which PR are you talking about?

> * `net.to('cuda')`
>   OR
> * `net.set_params(device='cuda')`

I think for me the alternative would be either `net.to(device)` or having a setter for device so that `x.device = device` also works (might not be trivial to change).

> Can you expand on that in reference to the second option? How would that be different from any other set_params call?

I think what is meant is that for other parameters, you just do `self.foo = bar` (after resolving the sklearn dispatching logic) but here you need to do a bunch of more operations when the new device is different from the old one.

Another argument for using `net.to` instead of `net.set_params` is that device handling is clearly in the pytorch domain and not in the sklearn domain. From this point of view, I would also prefer `net.to`.

Maybe we could add a warning if someone uses `net.set_params(device=device)`, as this has almost certainly not the intended effect. But that still leaves open the potential error of setting `net.device = device`.

**Comment by githubnemo** [comment]
> > We already have this addition in one of the open PRs
> 
> Which PR are you talking about?

#557 introduces a `to_device` utility function.


> > * `net.to('cuda')`
> >   OR
> > * `net.set_params(device='cuda')`
> 
> I think for me the alternative would be either `net.to(device)` or having a setter for device so that `x.device = device` also works (might not be trivial to change).
> 
> > Can you expand on that in reference to the second option? How would that be different from any other set_params call?
> 
> I think what is meant is that for other parameters, you just do `self.foo = bar` (after resolving the sklearn dispatching logic) but here you need to do a bunch of more operations when the new device is different from the old one.

Yeah, that is the point essentially. I was considering the fact that some parameters involve re-initalization and some do not. `device` might be considered a parameter that requires re-initialization but in most cases the intent is another one (i.e. moving a trained net to cpu). This is very similar to the learning rate parameter which is also excempt from re-init. I am wondering if it is a good idea to mix parameters that re-initialize and those which don't over one interace (`set_params`).

> Another argument for using `net.to` instead of `net.set_params` is that device handling is clearly in the pytorch domain and not in the sklearn domain. From this point of view, I would also prefer `net.to`.
> 
> Maybe we could add a warning if someone uses `net.set_params(device=device)`, as this has almost certainly not the intended effect. But that still leaves open the potential error of setting `net.device = device`.

To me it seems at the moment that `set_params` should re-initialize when `device` is supplies and that there should be a way to move a trained net to cpu, e.g. `to()` which is separate.

**Comment by stanbiryukov** [comment]
Appreciate the thoughtful discussion. Any traction on agreed next steps or are we waiting on other PRs?

**Comment by BenjaminBossan** [comment]
I don't believe there are any blocking PRs, someone just needs to sit down and implement what we agreed upon. When doing this, some new issues may arise or some will solve themselves.

I might try my hand on it this weekend if I find time, but I'd also be happy to let someone else step in :)

**Comment by stanbiryukov** [comment]
Happy to help and review but you would have a better start than I, knowing the intricacies of the skorch software.

**Comment by ayaanhossain** [comment]
Since, this issue is open, just wanted to point out that one can add the following callback during a `skorch` model definition and it will train the model on GPU (if available) and move it to CPU for inference afterwards.

Define Freezer Callback
```
class FreezeMyMLP(skorch.callbacks.Callback):
    '''Move an MLP to cpu and trim the model for inference only.'''

    def on_train_end(self, net, **kwargs):
        '''Post training freezing of model'''

        # Freeze Model for Inference
        net.trim_for_prediction() # In future we'll use this model for inference only

        # Move Network to CPU
        net.set_params(device='cpu') # Ensures future model loads on cpu
        skorch.utils.to_device(net.module_, 'cpu') # Moves current model instance to cpu
```

Later Define the Model
```
net = NeuralNetClassifier(

    ..., # Usual arguments
    
    callbacks = [
      ..., # Other callbacks if any
      FreezeMyMLP(),
      ... # Other callbacks if any
      ]
      
    ... # Usual arguments
    
    device='cuda' if torch.cuda.is_available() else 'cpu', # doesn't matter honestly
)
```

Enjoy your model on CPU for inference only!
```
net.fit(X, y) # Potentially fit on GPU and due to FreezeMyMLP() model is now inference-only and on cpu
```
